### PR TITLE
[inductor] Remove identity from ops.scan

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122137
* #119430
* __->__ #119727

Currently scan has an `init` argument which must be the identity of the
combine function. This isn't strictly necessary if we are more careful about
keeping track of the first element and avoid combining it with anything.

This does additionally require that there are no active load masks, since we can't
do the `where_cond` any more. However, this shouldn't be possible anyway since
scans are always realized and only fused via the scheduler.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by peterbell10** [comment]
Looks like I'm running into a caching issue. The main triton kernel body has the same code, so it's using whichever combine function was compiled first, e.g.

```python
import torch

a = torch.randn(1, 100, device="cuda")

sum_res = torch.compile(torch.cumsum)(a, 1)
prod_res = torch.compile(torch.cumprod)(a, 1)
torch.allclose(sum_res, prod_res)

```

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -2367,9 +2367,17 @@ def _lift_helper(self, fn, num_args) -> str:
         cse = CSE(prefix="", suffix="")
         overrides = TritonOverrides(V.MockHandler())
 
+        # Build a name that changes depending on fn to workaround a triton bug
+        # where the combine_fn to reduce and scan is not hashed, and so different
+        # scan ops may collide in the triton cache.
+        # This is fixed with the latest triton pin, but not the triton-rocm pin.
+        helper_name = "_triton_helper_fn"
+
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:2375]: Rather than making the code more complex, we can probably just remove these workarounds and skip rocm really.

Comment (peterbell10) [reply torch/_inductor/codegen/triton.py:2375]: It's a pretty small change and also with backward compatibility for older triton verisons which is required because meta internal isn't kept in sync with oss.

File: torch/_inductor/triton_helpers.py [hunk 2]
```diff
@@ -269,15 +272,22 @@ def exclusive_scan_decoupled_lookback(
             flag = unpack_flag(pack, DTYPE_VALUE_AS_UINT)
 
         value = unpack_value(pack, DTYPE_VALUE, DTYPE_VALUE_AS_UINT)
-        exclusive_prefix = combine_fn(value, exclusive_prefix)
+        if prefix_valid:
+            exclusive_prefix = combine_fn(value, exclusive_prefix)
+        else:
+            exclusive_prefix = value
+            prefix_valid = True
```
Comment (lezcano) [inline torch/_inductor/triton_helpers.py:279]: Did you check whether this generates alright code in triton? If not, we can always unroll the first iteration.

**Review by lezcano — APPROVED** [review]
SG, just one minor point.

File: torch/_inductor/triton_helpers.py (hunk 2 above)
Comment (peterbell10) [reply torch/_inductor/triton_helpers.py:279]: I checked for cumsum and the llvm IR looks like this
```llvm

  %112 = phi float [ -0.000000e+00, %.lr.ph ], [ %121, %select.unfold ]
  ; ... snip ...

select.unfold:                                    ; preds = %115
  %extract.le = lshr i64 %117, 32
  %extract.t.le = trunc i64 %extract.le to i32
  %120 = bitcast i32 %extract.t.le to float, !dbg !44
  %121 = fadd float %112, %120, !dbg !45
  %122 = icmp ne i32 %118, 2, !dbg !46
  %123 = add nsw i32 %111, -1, !dbg !47
  %124 = icmp sgt i32 %111, 0, !dbg !39
  %125 = select i1 %122, i1 %124, i1 false, !dbg !39
  br i1 %125, label %110, label %select.unfold._crit_edge, !dbg !39
```

Where we can see `%112` is initialized to `0.0e+0` and unconditionally updated with `fadd float %112, %120` inside the loop body. We can also see that updating `test_target` below has been transformed from an if-else to a `select` operation.

For `cumprod` I see an almost identical kernel, except that `%121` is initialized to `1.0e+0` and `fadd` becomes `fmul`. So it seems that llvm was able to infer the correct identity and setup the loop appropriately.

Comment (lezcano) [reply torch/_inductor/triton_helpers.py:279]: That's quite cool actually. I wonder whether it's able to figure out what to do in more complex cases, but this is certainly good enough for now.
